modules/ding_innskraning/ding_innskraning.module

```diff
@@ -0,0 +1,780 @@
+ * Try to login user by SSN number.
+ */
+function _ding_innskraning_activate_user_by_ssn($ssn) {
+  $auth_res = array(
+    'success' => TRUE,
+    'creds' => array(
+      'name' => $ssn,
+      // Random password. Shouldn't be used, but we'll make it un-guessable
+      // for good measure, in case someone messes up.
+      'pass' => md5(drupal_random_bytes(32) . 'no_password_for_the_wicked'),
+      // Note that user was logged in externally.
+      'external-auth' => TRUE,
+    ),
+    'user' => array(
+      'mail' => '',
+      'blocked' => FALSE,
+      'data' => array(
+        'display_name' => '',
+      ),
+    ),
+  );
+
+  /*
+   * We call two private ding_user functions to do the heavy lifting. Sadly,
+   * nobody has re-factored ding_user to have a simple login function, so we
+   * have to duplicate its work here.
+   */
+
+  // Create new account in Drupal and if one exists update it.
+  $account = _ding_user_create_account($ssn, $auth_res);
+
+  // Check that a profile exists for the user and if not create one.
+  _ding_user_create_profile2($account);
+
+  $form_state['uid'] = $account->uid;
+  user_login_submit(array(), $form_state);
+  return TRUE;
+}
+
+/**
+ * Verify the signatures and data from island.is.
+ */
+function _ding_innskraning_verify($xmlDoc) {
+  global $base_url;
+
+  // Load library.
+  libraries_load('xmlseclibs');
+
+  // Check SAML version.
+  if ($xmlDoc->documentElement->getAttribute('Version') != '2.0') {
+    throw new Exception('Unsupported SAML version');
+  }
+  // Check ID attribute.
+  if (!$xmlDoc->documentElement->hasAttribute('ID')) {
+    throw new Exception('Missing ID attribute on SAML Response');
+  }
+
+  // Check SAML Response status.
+  $status = $xmlDoc->getElementsByTagName('Status');
+
+  if ($status->length == 0) {
+    throw new Exception('Missing Status on response');
+  }
+
+  $statusCode = $status->item(0)->getElementsByTagName('StatusCode');
+
+  if ($statusCode->length == 0) {
+    throw new Exception('Missing Status Code on response');
+  }
+
+  $code = $statusCode->item(0)->getAttribute('Value');
+
+  if ($code != 'urn:oasis:names:tc:SAML:2.0:status:Success') {
+    throw new Exception('User was not authenticated with Island.is');
+  }
+
+  // Check number of Assertions.
+  $singleAssertion = $xmlDoc->getElementsByTagName('Assertion');
+  if ($singleAssertion->length != 1) {
+    throw new Exception('SAML Response must contain 1 assertion');
+  }
+
+  // Validate XML file structure.
+  $schemaFile = libraries_get_path('php-saml') . '/lib/Saml2/schemas/saml-schema-protocol-2.0.xsd';
+  $oldEntityLoader = libxml_disable_entity_loader(FALSE);
+  $res = $xmlDoc->schemaValidate($schemaFile);
+  libxml_disable_entity_loader($oldEntityLoader);
+  if (!$res) {
+    $xmlErrors = libxml_get_errors();
+    watchdog('ding_innskraning', 'Error validating the metadata: @vars', array('@vars' => var_export($xmlErrors, TRUE)));
+    throw new Exception('Error validating the metadata');
+  }
```

Comment:
It seems as if much of this validation could be handled by [`Response` for `php-saml`](https://github.com/onelogin/php-saml/blob/master/lib/Saml2/Response.php#L100).

Reply:
Yeah, it would seem so...